registry/storage/revisionstore.go

```diff
@@ -79,12 +90,30 @@ func (rs *revisionStore) get(revision digest.Digest) (*manifest.SignedManifest,
 // put stores the manifest in the repository, if not already present. Any
 // updated signatures will be stored, as well.
 func (rs *revisionStore) put(sm *manifest.SignedManifest) (digest.Digest, error) {
+	ctx := rs.repository.ctx
 	// Resolve the payload in the manifest.
 	payload, err := sm.Payload()
 	if err != nil {
 		return "", err
 	}
 
+	// This put may be restoring a previously deleted manifest
+	// If so, remove the tombstone
+	dgst, err := digest.FromBytes(payload)
+	if err != nil {
+		return "", err
+	}
+
+	tombstoneExists, err := rs.tomb.tombstoneExists(ctx, rs.Name(), dgst)
```

Comment:
Maybe move the exists check inside deleteTombstone? I feel like it would make this a little cleaner and now we don't have to code up an exists check everywhere we want to try and delete a tombstone. If the tombstone doesn't exist, I'd expect delete to return a nil err (same "final state is as desired" idea that I mentioned above)